src/storage/txn/commands.rs

```diff
@@ -317,6 +325,39 @@ macro_rules! tag {
     };
 }
 
+macro_rules! write_bytes {
+    ($field: ident) => {
+        fn write_bytes(&self) -> usize {
+            self.$field.as_encoded().len()
+        }
+    };
+    ($field: ident, multiple) => {
+        fn write_bytes(&self) -> usize {
+            self.$field.iter().map(|x| x.as_encoded().len()).sum()
+        }
+    };
+}
+
+macro_rules! gen_lock {
+    ($field: ident) => {
+        fn gen_lock(&self, latches: &Latches) -> latch::Lock {
+            latches.gen_lock(&[&self.$field])
+        }
+    };
+    ($field: ident: multiple) => {
+        fn gen_lock(&self, latches: &Latches) -> latch::Lock {
+            latches.gen_lock(&self.$field)
+        }
+    };
+    ($field: ident: multiple$transform: tt) => {
+        fn gen_lock(&self, latches: &Latches) -> latch::Lock {
+            #![allow(unused_parens)]
+            let keys: Vec<&Key> = self.$field.iter().map($transform).collect();
```

Comment:
How about let `latches.gen_lock` receive an iterator so we don't need to `collect` here, which requires allocation?

Reply:
I'll open another pr for this.